manuals/vhdiinfo.1
.Dd June 23, 2026
.Dt VHDIINFO 1
.Os
.Sh NAME
.Nm vhdiinfo
.Nd determines information about a Virtual Hard Disk (VHD) image file
.Sh SYNOPSIS
.Nm vhdiinfo
.Op Fl hvV
.Ar source
.Sh DESCRIPTION
.Nm vhdiinfo
is a utility to determine information about a Virtual Hard Disk (VHD) image file
.Pp
.Nm vhdiinfo
is part of the
.Nm libvhdi
package.
.Nm libvhdi
is a library to access the Virtual Hard Disk (VHD) image format
.Pp
.Ar source
is the source image.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl h
shows this help
.It Fl v
verbose output to stderr
.It Fl V
print version
.El
.Sh ENVIRONMENT
None
.Sh FILES
None
.Sh EXAMPLES
.Bd -literal
# vhdiinfo differential.vhd
vhdiinfo 20130512
.sp
Virtual Hard Disk (VHD) image information:
	Format:			1.0
	Disk type:		Differential
	Media size:		536870912 bytes
	Identifier:		5577d188-a753-4a80-800a-c627f764c304
	Parent identifier:	44421587-1b53-4972-9ceb-9a31e1618e5b
	Parent filename: 	dynamic.vhd
.sp
.Ed
.Sh DIAGNOSTICS
Errors, verbose and debug output are printed to stderr when verbose output \
\-v is enabled.
Verbose and debug output are only printed when enabled at compilation.
.Sh AUTHORS
.An Joachim Metz <joachim.metz@gmail.com>
.Sh BUGS
Please report bugs of any kind on the project issue tracker: \
https://github.com/libyal/libvhdi/issues
.Sh COPYRIGHT
Copyright (C) 2012-2026, Joachim Metz <joachim.metz@gmail.com>.
.sp
This is free software; see the source for copying conditions.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A \
PARTICULAR PURPOSE.


ChangeLog
TODO
* add VHDX block state sanity checks
* move parent identifier and filename to IO handle ?
* validated checksums
* add VHDX log (journal) support
* debug print timestamps
* improve test coverage

Tests:
* Test concurrent read from diff image and parent
* Add verify test

* debug
  - print creator application
  - print creator host system
  - print disk geometry
* what about geometry and disk size calculation
* read backup file footer of dynamic disk

20130922
* updated msvscpp files
* removed libfvalue
* see `git log' for more recent change log

20130916
* updated dependencies
* worked on libcthreads build support
* code clean up

20130831
* updated dependencies

20130514
* changes in parent functions

20130512
* updated dependencies
* changes to macosx files
* worked on Python bindings
* worked on debug output
* worked on differential image support
* worked on Dokan support

20130501
* vhdimount: worked on differential images support

20121209
* changes for fdata refactor

20121208
* changes for fdata refactor

20121202
* changes for fdata refactor

20121201
* updated dependencies

20120929
* updated dependencies

20120922
* code clean up

20120920
* updated dependencies

20120918
* added support for fixed disk image

20120917
* fix for debian rules file

20120914
* fix for bitmap size, needs to be a multitude of sector size

20120913
* initial version



vhditools/vhdiinfo.c
/*
 * Shows information obtained from a Virtual Hard Disk (VHD) image file.
 *
 * Copyright (C) 2012-2026, Joachim Metz <joachim.metz@gmail.com>
 *
 * Refer to AUTHORS for acknowledgements.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#include <common.h>
#include <file_stream.h>
#include <system_string.h>
#include <types.h>

#if defined( HAVE_FCNTL_H ) || defined( WINAPI )
#include <fcntl.h>
#endif

#if defined( HAVE_IO_H ) || defined( WINAPI )
#include <io.h>
#endif

#if defined( HAVE_STDLIB_H ) || defined( WINAPI )
#include <stdlib.h>
#endif

#if defined( HAVE_UNISTD_H )
#include <unistd.h>
#endif

#include "info_handle.h"
#include "vhditools_getopt.h"
#include "vhditools_libcerror.h"
#include "vhditools_libclocale.h"
#include "vhditools_libcnotify.h"
#include "vhditools_libvhdi.h"
#include "vhditools_output.h"
#include "vhditools_signal.h"
#include "vhditools_unused.h"

info_handle_t *vhdiinfo_info_handle = NULL;
int vhdiinfo_abort                  = 0;

/* Signal handler for vhdiinfo
 */
void vhdiinfo_signal_handler(
      vhditools_signal_t signal VHDITOOLS_ATTRIBUTE_UNUSED )
{
	libcerror_error_t *error = NULL;
	static char *function    = "vhdiinfo_signal_handler";

	VHDITOOLS_UNREFERENCED_PARAMETER( signal )

	vhdiinfo_abort = 1;

	if( vhdiinfo_info_handle != NULL )
	{
		if( info_handle_signal_abort(
		     vhdiinfo_info_handle,
		     &error ) != 1 )
		{
			libcnotify_printf(
			 "%s: unable to signal info handle to abort.\n",
			 function );

			libcnotify_print_error_backtrace(
			 error );
			libcerror_error_free(
			 &error );
		}
	}
	/* Force stdin to close otherwise any function reading it will remain blocked
	 */
#if defined( WINAPI ) && !defined( __CYGWIN__ )
	if( _close(
	     0 ) != 0 )
#else
	if( close(
	     0 ) != 0 )
#endif
	{
		libcnotify_printf(
		 "%s: unable to close stdin.\n",
		 function );
	}
}

/* The main program
 */
#if defined( HAVE_WIDE_SYSTEM_CHARACTER )
int wmain( int argc, wchar_t * const argv[] )
#else
int main( int argc, char * const argv[] )
#endif
{
	const char *description = \
		"Use vhdiinfo to determine information about a Virtual Hard Disk (VHD) image file.";

	vhditools_option_t options[ ] = {
		{ 'h', NULL, "shows this help" },
		{ 'v', NULL, "verbose output to stderr" },
		{ 'V', NULL, "print version" },
		{ 0, "source", "the source image" },
	};
	system_character_t options_string[ 32 ];

	libvhdi_error_t *error     = NULL;
	system_character_t *source = NULL;
	char *program              = "vhdiinfo";
	system_integer_t option    = 0;
	int number_of_options      = (int) ( sizeof( options ) / sizeof( vhditools_option_t ) );
	int verbose                = 0;

#if defined( __MINGW32__ ) && defined( HAVE_MINGW_BINMODE )
	_setmode( _fileno( stdout ), _O_BINARY );
	_setmode( _fileno( stderr ), _O_BINARY );
#endif

	libcnotify_stream_set(
	 stderr,
	 NULL );
	libcnotify_verbose_set(
	 1 );

	if( libclocale_initialize(
	     "vhditools",
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize locale values.\n" );

		goto on_error;
	}
	if( vhditools_output_initialize(
	     _IONBF,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize output settings.\n" );

		goto on_error;
	}
	vhditools_output_version_fprint(
	 stdout,
	 program );

	if( vhditools_getopt_get_options_string(
	     options,
	     number_of_options,
	     options_string,
	     32 ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to determine options string.\n" );

		goto on_error;
	}
	while( ( option = vhditools_getopt(
	                   argc,
	                   argv,
	                   options_string ) ) != (system_integer_t) -1 )
	{
		switch( option )
		{
			case (system_integer_t) '?':
			default:
				fprintf(
				 stderr,
				 "Invalid argument: %" PRIs_SYSTEM "\n",
				 argv[ optind - 1 ] );

				vhditools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_FAILURE );

			case (system_integer_t) 'h':
				vhditools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_SUCCESS );

			case (system_integer_t) 'v':
				verbose = 1;

				break;

			case (system_integer_t) 'V':
				vhditools_output_copyright_fprint(
				 stdout );

				return( EXIT_SUCCESS );
		}
	}
	if( optind == argc )
	{
		fprintf(
		 stderr,
		 "Missing source file.\n" );

		vhditools_getopt_usage_fprint(
		 stdout,
		 program,
		 description,
		 options,
		 number_of_options );

		return( EXIT_FAILURE );
	}
	source = argv[ optind ];

	libcnotify_verbose_set(
	 verbose );
	libvhdi_notify_set_stream(
	 stderr,
	 NULL );
	libvhdi_notify_set_verbose(
	 verbose );

	if( info_handle_initialize(
	     &vhdiinfo_info_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize info handle.\n" );

		goto on_error;
	}
	if( info_handle_open_input(
	     vhdiinfo_info_handle,
	     source,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to open source file.\n" );

		goto on_error;
	}
	if( info_handle_image_fprint(
	     vhdiinfo_info_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to print image information.\n" );

		goto on_error;
	}
	if( info_handle_close(
	     vhdiinfo_info_handle,
	     &error ) != 0 )
	{
		fprintf(
		 stderr,
		 "Unable to close info handle.\n" );

		goto on_error;
	}
	if( info_handle_free(
	     &vhdiinfo_info_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to free info handle.\n" );

		goto on_error;
	}
	return( EXIT_SUCCESS );

on_error:
	if( error != NULL )
	{
		libcnotify_print_error_backtrace(
		 error );
		libcerror_error_free(
		 &error );
	}
	if( vhdiinfo_info_handle != NULL )
	{
		info_handle_free(
		 &vhdiinfo_info_handle,
		 NULL );
	}
	return( EXIT_FAILURE );
}

